src/python/pants/engine/fs.py

```diff
@@ -155,6 +155,10 @@ class MergedDirectories(datatype([('directories', tuple)])):
   pass
 
 
+class PrefixStrippedDirectory(datatype([('directory_digest', Digest), ('prefix', text_type)])):
```

Comment:
Should `prefix` have its own `datatype`? I'm thinking we'll want this in my followup PR to add source root rule.

We likely want to keep it generic and call it `Prefix`, but maybe call it a `SourceRoot`? The information I think we're going to care about is the actual prefix path (`text_type`) and maybe the language? Or should language be embedded in the prefix? 

Example source root we need to support: `src/python` and `contrib/go/src/python`. With this example, I think just having it be the path portion works best; I'm not convinced every single source root will even have a language, such as for monolingual projects.

Reply:
This level of abstraction doesn't involve source roots at all; we may want a `Path` wrapper at some point instead of `text_type` (though that makes the Rust side a little more ugly, as it needs to do multiple layers of projections, rather than just one).

Certainly for a rule you may introduce on top of this one, having a `SourceRoot` wrapper there could make sense, but this layer should probably just talk about paths (which are currently strings).

Reply:
Agreed that the abstraction `SourceRoot` is wrong here.

A `Path` wrapper sounds highly correct, though, even though it increases the ceremony of everything. It's much easier to read a rule that's `@rule(PrefixStrippedDirectory, [DirectoryDigest, Prefix])` than `@rule(PrefixStrippedDirectory, [DirectoryDigest, text_type])`.

Per https://github.com/pantsbuild/pants/blob/master/src/python/pants/engine/README.md#datatypes, could we introduce this datatype to `fs.py` as a simple wrapper around `text_type`, or will that require changes to Rust? In type theory land, we want a type _alias_, not a new type.

Reply:
If we introduce a wrapper like this:
```python
class Path(datatype([('underlying_path', text_type)]):
  pass
```

then in rust every place we currently:
```rust
let prefix = externs::project_str(&request, "prefix");
```

we'd need to instead:
```rust
let prefix_wrapper = externs::project_ignoring_type(&request, "prefix");
let prefix = externs::project_str(&prefix_wrapper, "underlying_path");
```

It would also mean that on the rust side, every time we want to return a path, we'd need to call an additional python function; i.e.
```rust
let value = externs::store_utf8(s);
```

would become:
```rust
let string_value = externs::store_utf8(s);
let wrapper_value = externs::unsafe_call(&core.types.construct_path_wrapper, &[string_value]);
```

I suspect that the runtime overhead of doing this would be noticeable, but probably not hugely significant.


We could do the less involved thing of just having a `PathWrapper = text_type` and referring to the thing as `PathWrapper` instead of `text_type`, but we'd not get any value other than documentation here; nothing would enforce that the correct thing is used.

Reply:
> We could do the less involved thing of just having a `PathWrapper = text_type` and referring to the thing as `PathWrapper` instead of `text_type`, but we'd not get any value other than documentation here; nothing would enforce that the correct thing is used.

I think there is a lot of value in this, and that this is a great place to start. I'd love to see this in this PR if possible, so that the class definition becomes:

```
class PrefixStrippedDirectory(datatype([('directory_digest', Digest), ('prefix', Path)])):
```

Also I think we can call it `Path` instead of `PathWrapper`. It would make it easier on the Python side of things to down the road promote the alias to a full datatype, if we choose to do so. And, it's not really a wrapper, but simply an alias.

Reply:
There's already a type called `pants.engine.fs.Path` which is used for something else :(

Reply:
Are you planning to merge https://github.com/pantsbuild/pants/pull/7701 before this one, so that you can use the name `Path`?

Not necessary to block if you prefer to get it in before you leave for the weekend, although I do prefer not using `text_type` in the dataclass if possible. I could maybe fix in a followup.

Reply:
IMO, the outer wrapper provides enough type information here.